Explicitely specify cmp instructions to be on 64bits for ppc64le

tfauck opened the pull request:
Pointer to function in f_expand is 64bits on ppc64le
    For ppc64le unpack may result in +1 or +3 offset

Signed-off-by: Thierry Fauck <tfauck@free.fr>

	modified:   stub/scripts/xstrip.py
	modified:   stub/src/arch/powerpc/32/bxx.S
	modified:   stub/src/arch/powerpc/64le/bxx.S
	modified:   stub/src/arch/powerpc/64le/lzma_d.S
	modified:   stub/src/arch/powerpc/64le/nrv2b_d.S
	modified:   stub/src/arch/powerpc/64le/nrv2d_d.S
	modified:   stub/src/arch/powerpc/64le/nrv2e_d.S
	modified:   stub/src/powerpc64le-darwin.dylib-entry.S
	modified:   stub/src/powerpc64le-darwin.macho-entry.S
	modified:   stub/src/powerpc64le-darwin.macho-fold.S
	modified:   stub/src/powerpc64le-linux.elf-entry.S
	modified:   stub/src/powerpc64le-linux.elf-fold.S
	modified:   stub/src/powerpc64le-linux.elf-main.c
	modified:   stub/src/powerpc64le-linux.kernel.vmlinux.S

markus-oberhumer commented:
X=rebuild stubs fails - please see the Travis log.

tfauck commented:
On 13/01/2017 01:34, Markus F.X.J. Oberhumer wrote:
>
> X=rebuild stubs fails - please see the Travis log.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/upx/upx/pull/51#issuecomment-272327909>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ANQxU59tU8dursmy7GJohwRLkylvPV4hks5rRsakgaJpZM4Lhzs0>.
>

Hello,

I realised that on build #16 - So rebuilt them and #17 passed this
night. Let me know if I missed something

Just for your info, now on Ubuntu it works with all compiler from gcc-4
to 6.

I will have a new patch for Fedora and Debian later on when I will
figure out how to get the info from libc..

Thanks

-- 

__________________________________________________________________________
thf - Thierry Fauck - tfauck@free.fr>
/pubkey: 4096R/FCC181CE/
/fingerprint: 5CCF 6B82 DE4E E72A A40B B63E A153 BF4F FCC1 81CE/

jreiser commented on src/stub/scripts/xstrip.py:
```diff
@@ -161,7 +161,8 @@ def do_file(fn):
         e_shnum, e_shstrndx = struct.unpack("<HH", eh[60:64])
-        assert e_shstrndx + 3 == e_shnum
+        if ( e_shstrndx + 1 != e_shnum  and  e_shstrndx + 3 != e_shnum ):
+            assert e_shnum == e_shstrndx
```
Please comment on why there could be a choice between +1 and +3.

tfauck replied:
On ubuntu 16.04 I get 
python ./../../src/stub/scripts/xstrip.py --with-dump=tmp/powerpc64le-darwin.dylib-entry.bin.dump --create-bindump=tmp/powerpc64le-darwin.dylib-entry.bin.bindump tmp/powerpc64le-darwin.dylib-entry.bin
e_shnum= 22 e_shstrndx = 19
but on Ubuntu 17.04, debian and fedora I have 
e_shnum= 22 e_shstrndx = 21

I agree it must be a separated commit, but I need to understand what it makes that difference. I have checked it is not related to gcc version.

jreiser commented on src/stub/src/arch/powerpc/64le/bxx.S:
```diff
@@ -36,10 +36,10 @@ ppcbxx:  # (*f_unf)(xo->buf, out_len, h.b_cto8, h.b_ftid);
 #define ptr0 a4
 
-        cmpli cr0,ftid,0xd0;       bnelr- cr0  # if (0xd0!=ftid)   return;
+        cmpldi cr0,ftid,0xd0;       bnelr- cr0  # if (0xd0!=ftid)   return;
```
ftid is a 32-bit quantity.  Why is a 64-bit cmpldi being used?

jreiser commented on src/stub/src/arch/powerpc/64le/bxx.S:
```diff
@@ -36,10 +36,10 @@ ppcbxx:  # (*f_unf)(xo->buf, out_len, h.b_cto8, h.b_ftid);
         rlwinm. len,len,32-2,2,31; beqlr- cr0  # if (0==(len>>=2)) return;
         lis r0,-(~0<<(32-16- (2+6+ W_CTO)))  # limit in 32-bit words
-        cmpl cr0,len,r0
+        cmpld cr0,len,r0
```
len is a 32-bit quantify.

jreiser commented on src/stub/src/arch/powerpc/64le/bxx.S:
```diff
@@ -58,7 +58,7 @@ L10:
         lwzu t0,4(ptr)  # t0= *++ptr;
         rlwinm t1,t0,6+W_CTO,32-(6+W_CTO),31  # t1= top (6+W_CTO) bits of t0
-        cmpl cr0,t1,cto8; beq- cr0,L10  # unconditional branch marked with cto8; unlikely
+        cmpld cr0,t1,cto8; beq- cr0,L10  # unconditional branch marked with cto8; unlikely
```
cto8 is 32 bits.

jreiser commented on src/stub/src/arch/powerpc/64le/lzma_d.S:
```diff
@@ -44,7 +44,7 @@ retaddr = 6*8  // (sp,cr,pc, xx,yy,zz) save area per calling convention
 
 #define M_LZMA          14
-        cmpli cr0,meth,M_LZMA
+        cmpldi cr0,meth,M_LZMA
```
meth is 32 bits.

jreiser commented on src/stub/src/arch/powerpc/64le/nrv2b_d.S:
```diff
@@ -33,7 +33,7 @@
         mflr t3  // return address
 
-        cmpli cr0,meth,M_NRV2B_LE32
+        cmpldi cr0,meth,M_NRV2B_LE32
```
meth is 32 bits.

jreiser commented on src/stub/src/arch/powerpc/64le/nrv2d_d.S:
```diff
@@ -33,7 +33,7 @@
         mflr t3  // return address
 
-        cmpli cr0,meth,M_NRV2D_LE32
+        cmpldi cr0,meth,M_NRV2D_LE32
```
meth is 32 bits.

jreiser commented on src/stub/src/arch/powerpc/64le/nrv2e_d.S:
```diff
@@ -33,7 +33,7 @@
         mflr t3  // return address
 
-        cmpli cr0,meth,M_NRV2E_LE32
+        cmpldi cr0,meth,M_NRV2E_LE32
```
meth is 32 bits.

jreiser commented on src/stub/src/powerpc64le-darwin.macho-fold.S:
```diff
@@ -71,7 +71,7 @@ L90:
         subf a0,a1,r31  # &l_info
         lwz a3,sz_unc+sz_p_info+sz_l_info(a0)  # sz_mach_headers
-        cmpli 0,a3,2048; bgt L100; li a3,2048  # at least 2 KiB for /usr/lib/dyld
+        cmpldi 0,a3,2048; bgt L100; li a3,2048  # at least 2 KiB for /usr/lib/dyld
```
a3 is 32 bits ('lwz' at line 73)

jreiser commented on src/stub/src/powerpc64le-linux.elf-fold.S:
```diff
@@ -95,20 +95,27 @@ L90:
-        std   12,-27984(r31)  // required : Inconsistency detected by ld.so: rtld.c:
+
+/* required : Inconsistency detected by ld.so: rtld.c: */
+        addi  12,r31,-27828  // Ubuntu: r31=0x57e00 -> 0x5114c
```
Use hexadecimal.

tfauck replied:
That's where I am working right now trying to figure how to get these value from the ld64.so - Comparing process with /lib64/ld64.so.2 binary and execution of upx_generated_binary I found that on Ubuntu at least 2 areas need to be initialized or cleared. I need  to find proper way to initalize the areas.

jreiser commented on src/stub/src/powerpc64le-linux.elf-fold.S:
```diff
@@ -95,20 +95,27 @@ L90:
 
         li    12,0
-        std   12,-28192(r31)    // p 0x...d7e00-0x...d0fe0
+        std   12,-28192(r31) // Ubuntu: in 0x50fe0
```
Please comment on what is going on here.  Why the magic addresses?  When *WILL* they change?  How were the values discovered? Etc.

jreiser reviewed (changes requested):
Perhaps the changes in the diff are necessary, but we must have documentation about _why_ and _how_.

tfauck commented:
Thanks for your comments - To complete some of my answers, for the cmp questions I need to review them but on 64 bits in order to have the overflow bits set I need to use 64 bits registers so in the process I load 32bits (as in powerpc) but then I shift data explaining that some of the cmp must be 64bits. Initialy I didn't specify the 32bits or 64bits comparisons, but on one OS I got errors asking me to specify that.

tfauck commented:
Did review the cmpld comments above - you are right - these are 32bits comparisons - going to redo my patch and separate the commits. thx

markus-oberhumer commented:
What's the current status of this?

tfauck commented:
On 27/01/2017 08:50, Markus F.X.J. Oberhumer wrote:
>
> What's the current status of this?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/upx/upx/pull/51#issuecomment-275607797>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ANQxU5yYyoRlsVNI051MSDT5zCmKD-JWks5rWaHBgaJpZM4Lhzs0>.
>
Sorry for such delay,

Static displacement is currently only valid for Ubuntu 16.04 up to 17.04
using gcc 4.9 up to gcc 6.0

I found some relation between these values and some variables in ld.so

but this is only true for Ubuntu.

On Debian and Fedora It sounds I need to clear pages allocated for .bss
(suspected to do that in do_xmap() in

"if (addr < haddr) { // need pages for .bss " but it is overwritten and
I haven't find exactly with what.

SO I don't know really how to handle that properly - I may come with a
version for Ubuntu for now if you need it

What do you suggest ?

Thanks

Thierry

#-------------------------------Ubuntu ld64.so displacements and
variables ----------------------------------

// r2   +0x57e00

// 0x510b0 <alloc_ptr>
// 0x51100 <rtld_errno>
// 0x51108 <r_debug>
// 0x50fd8 <_dl_rtld_libname>
// .bss=data.10608 @ 0x50fc0
//                         //    end of .bss=0x40fc0+0x170=0x51130
//       addi  12,r31,128-27984  // get rtld_errno+40 address
128-27984=-27856=-0x6cd0
//                               // 0x57e00-0x6cd0=0x51130 end of .bss
// set in alloc_ptr @ 0x57e00-27984=0x57e00-0x6d50=0x510b0

// Inconsistency detected by ld.so: rtld.c: 1154:
//      dl_main: Assertion `GL(dl_rtld_map).l_libname->next == NULL' failed

// Ubuntu: in _dl_rtld_libname+8=0x50fe0=.bss+0x20

jreiser commented:
Please put the changes for "cmpl ==> cmpld", and _only_ those changes, into a separate commit; it will be approved and merged.  Then we can work on the issue of corrupted address space with ld.so and .bss.

tfauck committed:
Explicitely specify cmp instructions to be 32 bits or 64bits on powerpc

Signed-off-by: Thierry Fauck <tfauck@free.fr>

	modified:   stub/src/arch/powerpc/32/bxx.S
	modified:   stub/src/arch/powerpc/64le/bxx.S
	modified:   stub/src/arch/powerpc/64le/lzma_d.S
	modified:   stub/src/arch/powerpc/64le/nrv2b_d.S
	modified:   stub/src/arch/powerpc/64le/nrv2d_d.S
	modified:   stub/src/arch/powerpc/64le/nrv2e_d.S
	modified:   stub/src/powerpc64le-darwin.dylib-entry.S
	modified:   stub/src/powerpc64le-darwin.macho-entry.S
	modified:   stub/src/powerpc64le-darwin.macho-fold.S
	modified:   stub/src/powerpc64le-linux.elf-entry.S
	modified:   stub/src/powerpc64le-linux.elf-fold.S
	modified:   stub/src/powerpc64le-linux.kernel.vmlinux.S

tfauck commented:
Recreating patch

jreiser commented:
@tfauck  It seems to me that we require 64-bit compares whenever 'dst', 'dst0', or 'sp' are involved:
```
$ git diff $(find . -name powerpc64le'*'.S)
diff --git a/src/stub/src/powerpc64le-darwin.dylib-entry.S b/src/stub/src/powerpc64le-darwin.dylib-entry.S
index 4242e5b..2f183ed 100644
--- a/src/stub/src/powerpc64le-darwin.dylib-entry.S
+++ b/src/stub/src/powerpc64le-darwin.dylib-entry.S
@@ -101,7 +101,7 @@ CACHELINE=32
         ori dst0,dst0,-1+ CACHELINE  // highest addr on cache line
 cfl_nrv:
         dcbst  0,dst0  // initiate store (modified) cacheline to memory
-        cmpl cr0,0,dst0,dst  // did we cover the highest-addressed byte?
+        cmpl cr0,1,dst0,dst  // did we cover the highest-addressed byte?
         icbi   0,dst0  // discard instructions from cacheline
         addi     dst0,dst0,CACHELINE  // highest addr on next line
         blt  cr0,cfl_nrv  // not done yet
diff --git a/src/stub/src/powerpc64le-darwin.macho-entry.S b/src/stub/src/powerpc64le-darwin.macho-entry.S
index 8d38bb9..7260424 100644
--- a/src/stub/src/powerpc64le-darwin.macho-entry.S
+++ b/src/stub/src/powerpc64le-darwin.macho-entry.S
@@ -87,7 +87,7 @@ CACHELINE=32
         ori dst0,dst0,-1+ CACHELINE  // highest addr on cache line
 cfl_nrv:
         dcbst  0,dst0  // initiate store (modified) cacheline to memory
-        cmpl cr0,0,dst0,dst  // did we cover the highest-addressed byte?
+        cmpl cr0,1,dst0,dst  // did we cover the highest-addressed byte?
         icbi   0,dst0  // discard instructions from cacheline
         addi     dst0,dst0,CACHELINE  // highest addr on next line
         blt  cr0,cfl_nrv  // not done yet
diff --git a/src/stub/src/powerpc64le-linux.elf-entry.S b/src/stub/src/powerpc64le-linux.elf-entry.S
index 3625be3..4a78bd5 100644
--- a/src/stub/src/powerpc64le-linux.elf-entry.S
+++ b/src/stub/src/powerpc64le-linux.elf-entry.S
@@ -114,7 +114,7 @@ CACHELINE=32
         ori dst0,dst0,-1+ CACHELINE  // highest addr on cache line
 cfl_nrv:
         dcbst  0,dst0  // initiate store (modified) cacheline to memory
-        cmpl cr0,0,dst0,dst  // did we cover the highest-addressed byte?
+        cmpl cr0,1,dst0,dst  // did we cover the highest-addressed byte?
         icbi   0,dst0  // discard instructions from cacheline
         addi     dst0,dst0,CACHELINE  // highest addr on next line
         blt  cr0,cfl_nrv  // not done yet
diff --git a/src/stub/src/powerpc64le-linux.kernel.vmlinux.S b/src/stub/src/powerpc64le-linux.kernel.vmlinux.S
index 7e4a57c..ce055ea 100644
--- a/src/stub/src/powerpc64le-linux.kernel.vmlinux.S
+++ b/src/stub/src/powerpc64le-linux.kernel.vmlinux.S
@@ -116,7 +116,7 @@ CACHELINE=32
         ori dst0,dst0,-1+ CACHELINE  // highest addr on cache line
 cfl_nrv:
         dcbst  0,dst0  // initiate store (modified) cacheline to memory
-        cmpl cr0,0,dst0,dst  // did we cover the highest-addressed byte?
+        cmpl cr0,1,dst0,dst  // did we cover the highest-addressed byte?
         icbi   0,dst0  // discard instructions from cacheline
         addi     dst0,dst0,CACHELINE  // highest addr on next line
         blt  cr0,cfl_nrv  // not done yet
```

tfauck commented:
Hello John,

I agree that dst0 and dst are supposed to 64bits addresses as they are
location addresses.
Currently I didn't get problem since entry address is located in the
32bits address space.
But if for some reasons it changes, then a problem will occur.
Thanks for pointing to that issue.
Regards
Thierry

On 02/02/2017 06:48, John Reiser wrote:
>
> @tfauck <https://github.com/tfauck> It seems to me that we require
> 64-bit compares whenever 'dst', 'dst0', or 'sp' are involved:
>
> |$ git diff $(find . -name powerpc64le'*'.S) diff --git
> a/src/stub/src/powerpc64le-darwin.dylib-entry.S
> b/src/stub/src/powerpc64le-darwin.dylib-entry.S index 4242e5b..2f183ed
> 100644 --- a/src/stub/src/powerpc64le-darwin.dylib-entry.S +++
> b/src/stub/src/powerpc64le-darwin.dylib-entry.S @@ -101,7 +101,7 @@
> CACHELINE=32 ori dst0,dst0,-1+ CACHELINE // highest addr on cache line
> cfl_nrv: dcbst 0,dst0 // initiate store (modified) cacheline to memory
> - cmpl cr0,0,dst0,dst // did we cover the highest-addressed byte? +
> cmpl cr0,1,dst0,dst // did we cover the highest-addressed byte? icbi
> 0,dst0 // discard instructions from cacheline addi dst0,dst0,CACHELINE
> // highest addr on next line blt cr0,cfl_nrv // not done yet diff
> --git a/src/stub/src/powerpc64le-darwin.macho-entry.S
> b/src/stub/src/powerpc64le-darwin.macho-entry.S index 8d38bb9..7260424
> 100644 --- a/src/stub/src/powerpc64le-darwin.macho-entry.S +++
> b/src/stub/src/powerpc64le-darwin.macho-entry.S @@ -87,7 +87,7 @@
> CACHELINE=32 ori dst0,dst0,-1+ CACHELINE // highest addr on cache line
> cfl_nrv: dcbst 0,dst0 // initiate store (modified) cacheline to memory
> - cmpl cr0,0,dst0,dst // did we cover the highest-addressed byte? +
> cmpl cr0,1,dst0,dst // did we cover the highest-addressed byte? icbi
> 0,dst0 // discard instructions from cacheline addi dst0,dst0,CACHELINE
> // highest addr on next line blt cr0,cfl_nrv // not done yet diff
> --git a/src/stub/src/powerpc64le-linux.elf-entry.S
> b/src/stub/src/powerpc64le-linux.elf-entry.S index 3625be3..4a78bd5
> 100644 --- a/src/stub/src/powerpc64le-linux.elf-entry.S +++
> b/src/stub/src/powerpc64le-linux.elf-entry.S @@ -114,7 +114,7 @@
> CACHELINE=32 ori dst0,dst0,-1+ CACHELINE // highest addr on cache line
> cfl_nrv: dcbst 0,dst0 // initiate store (modified) cacheline to memory
> - cmpl cr0,0,dst0,dst // did we cover the highest-addressed byte? +
> cmpl cr0,1,dst0,dst // did we cover the highest-addressed byte? icbi
> 0,dst0 // discard instructions from cacheline addi dst0,dst0,CACHELINE
> // highest addr on next line blt cr0,cfl_nrv // not done yet diff
> --git a/src/stub/src/powerpc64le-linux.kernel.vmlinux.S
> b/src/stub/src/powerpc64le-linux.kernel.vmlinux.S index
> 7e4a57c..ce055ea 100644 ---
> a/src/stub/src/powerpc64le-linux.kernel.vmlinux.S +++
> b/src/stub/src/powerpc64le-linux.kernel.vmlinux.S @@ -116,7 +116,7 @@
> CACHELINE=32 ori dst0,dst0,-1+ CACHELINE // highest addr on cache line
> cfl_nrv: dcbst 0,dst0 // initiate store (modified) cacheline to memory
> - cmpl cr0,0,dst0,dst // did we cover the highest-addressed byte? +
> cmpl cr0,1,dst0,dst // did we cover the highest-addressed byte? icbi
> 0,dst0 // discard instructions from cacheline addi dst0,dst0,CACHELINE
> // highest addr on next line blt cr0,cfl_nrv // not done yet |
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/upx/upx/pull/51#issuecomment-276874231>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ANQxU_hc0duuvKzPx3pl_OZBDyX1zf6uks5rYW4wgaJpZM4Lhzs0>.
>

markus-oberhumer commented:
Thierry,

please create a new pull request based on current origin/devel c4e8352 .

To avoid confusion I'm closing this issue #51.
